src/pieces/pawn.cpp

```diff
@@ -0,0 +1,236 @@
+                    }
+                }
+            }
+            i--;
+        } while (i>0);
+    }
+    //otherwise it can move only one square straight forward
+    if(this->color == Color::white)
+    {
+        test_coordinate = this->coordinate + DirectionOffset.at(Direction::up);
+        if(test_coordinate.is_valid())
+        {
+            test_piece = board.get_piece_at(test_coordinate);
+            if(test_piece == nullptr)
+            {
+                pseudo_moviments.push_back({this->coordinate,test_coordinate,false,false,false});
+            }
+        }
+    }else
+    {
+        test_coordinate = this->coordinate + DirectionOffset.at(Direction::down);
+        if(test_coordinate.is_valid())
+        {
+            test_piece = board.get_piece_at(test_coordinate);
+            if(test_piece == nullptr)
+            {
+                pseudo_moviments.push_back({this->coordinate,test_coordinate,false,false,false});
+            }
+        }
+    }
+    
+    //unlike other pieces, a pawn captures in front of it diagonally
+    if(this->color == Color::white)
+    {
+        test_coordinate = this->coordinate + DirectionOffset.at(Direction::left_up);
+        if(test_coordinate.is_valid())
+        {
+            test_piece = board.get_piece_at(test_coordinate);
+            if(test_piece != nullptr && this->color != test_piece->get_color() )
+            {
+                pseudo_moviments.push_back({this->coordinate,test_coordinate,false,false,false});
+            }
+        }
+        test_coordinate = this->coordinate + DirectionOffset.at(Direction::right_up);
+        if(test_coordinate.is_valid())
+        {
+            test_piece = board.get_piece_at(test_coordinate);
+            if(test_piece != nullptr && this->color != test_piece->get_color() )
+            {
+                pseudo_moviments.push_back({this->coordinate,test_coordinate,false,false,false});
+            }
+        }
+    }else
+    {
+        test_coordinate = this->coordinate + DirectionOffset.at(Direction::left_down);
+        if(test_coordinate.is_valid())
+        {
+            test_piece = board.get_piece_at(test_coordinate);
+            if(test_piece != nullptr && this->color != test_piece->get_color() )
+            {
+                pseudo_moviments.push_back({this->coordinate,test_coordinate,false,false,false});
+            }
+        }
+        test_coordinate = this->coordinate + DirectionOffset.at(Direction::right_down);
+        if(test_coordinate.is_valid())
+        {
+            test_piece = board.get_piece_at(test_coordinate);
+            if(test_piece != nullptr && this->color != test_piece->get_color() )
+            {
+                pseudo_moviments.push_back({this->coordinate,test_coordinate,false,false,false});
+            }
+        }
+    }
+    
+    //if a pawn arrives at the opposite side, can be promoted to another piece(every piece can be chosen)
```

Comment:
non capisco, la promozione non dovrebbe essere fatta nel turno corrente? In tal caso basta controllare se tra i pseudo_moves esiste un end coordinate che sia uguale a  0 o 7(dipende dal colore)